File: android_native/docx/src/main/kotlin/com/afluffywaffle/layuv/docx/model/ReadingPosition.kt
package com.afluffywaffle.layuv.docx.model

/** Mirror of lib/models/reading_position.dart. `fraction` is the 0.0–1.0 plain-text position. */
data class ReadingPosition(
    val mode: ReadingMode,
    val page: Int,
    val scrollOffset: Double,
    val fraction: Double = 0.0,
) {
    fun toMap(): Map<String, Any?> = linkedMapOf(
        "mode" to mode.name,
        "page" to page,
        "scrollOffset" to scrollOffset,
        "fraction" to fraction,
    )

    companion object {
        fun fromMap(map: Map<String, Any?>): ReadingPosition = ReadingPosition(
            mode = ReadingMode.byName(map["mode"] as String),
            page = (map["page"] as Number).toInt(),
            scrollOffset = (map["scrollOffset"] as Number).toDouble(),
            fraction = (map["fraction"] as? Number)?.toDouble() ?: 0.0,
        )
    }
}

File: macos_native/Packages/LeamhDocx/Sources/LeamhDocx/Model/ReadingPosition.swift
import Foundation

/// Mirror of ReadingPosition.kt / lib/models/reading_position.dart.
/// fraction is the 0.0–1.0 plain-text position.
public struct ReadingPosition: Equatable {
    public let mode: ReadingMode
    public let page: Int
    public let scrollOffset: Double
    public let fraction: Double

    public init(mode: ReadingMode, page: Int, scrollOffset: Double, fraction: Double = 0.0) {
        self.mode = mode
        self.page = page
        self.scrollOffset = scrollOffset
        self.fraction = fraction
    }

    func toMap() -> [String: Any?] {
        return [
            "mode": mode.rawValue,
            "page": page,
            "scrollOffset": scrollOffset,
            "fraction": fraction,
        ]
    }

    static func fromMap(_ map: [String: Any?]) -> ReadingPosition {
        ReadingPosition(
            mode: ReadingMode(rawValue: (map["mode"] as? String) ?? "") ?? .pageFlip,
            page: (map["page"] as? NSNumber)?.intValue ?? 0,
            scrollOffset: (map["scrollOffset"] as? NSNumber)?.doubleValue ?? 0.0,
            fraction: (map["fraction"] as? NSNumber)?.doubleValue ?? 0.0
        )
    }
}
